src/engine/renderer/glsl_source/cull_cp.glsl

```diff
@@ -56,59 +58,164 @@ struct GLIndirectCommand {
+		depthCoords.y = clamp( depthCoords.y, 0, int( ( u_ViewHeight >> level ) - 1 ) );
+		depthCoords.z = clamp( depthCoords.z, 0, int( ( u_ViewWidth >> level ) - 1 ) );
+		depthCoords.w = clamp( depthCoords.w, 0, int( ( u_ViewHeight >> level ) - 1 ) );
+
+		vec4 depthValues;
+		depthValues.x = texelFetch( depthImage, depthCoords.xy, level ).r;
+		depthValues.y = texelFetch( depthImage, depthCoords.zy, level ).r;
+		depthValues.z = texelFetch( depthImage, depthCoords.xw, level ).r;
+		depthValues.w = texelFetch( depthImage, depthCoords.zw, level ).r;
+
+		const float surfaceDepth = max( max( max( depthValues.x, depthValues.y ), depthValues.z ), depthValues.w );
+
+		culled = ( 1 + r_zNear / ( viewSpaceCenter.z + boundingSphere.radius ) ) > surfaceDepth;
+	}
+
+	return culled;
 }
 
 void ProcessSurfaceCommands( const in SurfaceDescriptor surface, const in bool enabled ) {
-    for( uint i = 0; i < MAX_SURFACE_COMMANDS; i++ ) {
-        const uint commandID = surface.surfaceCommandIDs[i];
-        if( commandID == 0 ) { // Reserved for no-command
-            return;
-        }
-        surfaceCommands[commandID + u_SurfaceCommandsOffset].enabled = enabled;
-    }
+	for( uint i = 0; i < MAX_SURFACE_COMMANDS; i++ ) {
+		const uint commandID = surface.surfaceCommandIDs[i];
+		if( commandID == 0 ) { // Reserved for no-command
+			return;
+		}
+		surfaceCommands[commandID + u_SurfaceCommandsOffset].enabled = enabled;
+	}
 }
 
 void main() {
-    const uint globalInvocationID = gl_GlobalInvocationID.z * gl_NumWorkGroups.x * gl_WorkGroupSize.x * gl_NumWorkGroups.y * gl_WorkGroupSize.y
-                             + gl_GlobalInvocationID.y * gl_NumWorkGroups.x * gl_WorkGroupSize.x
-                             + gl_GlobalInvocationID.x;
-    if( globalInvocationID >= u_TotalDrawSurfs ) {
-        return;
-    }
-    SurfaceDescriptor surface = surfaces[globalInvocationID];
-    bool culled = CullSurface( surface.boundingSphere );
-
-    ProcessSurfaceCommands( surface, !culled );
+	const uint globalGroupID = gl_WorkGroupID.z * gl_NumWorkGroups.x * gl_NumWorkGroups.y
+							 + gl_WorkGroupID.y * gl_NumWorkGroups.x
+							 + gl_WorkGroupID.x;
+	const uint globalInvocationID = gl_GlobalInvocationID.z * gl_NumWorkGroups.x * gl_WorkGroupSize.x * gl_NumWorkGroups.y * gl_WorkGroupSize.y
+								  + gl_GlobalInvocationID.y * gl_NumWorkGroups.x * gl_WorkGroupSize.x
+								  + gl_GlobalInvocationID.x;
+
+	// Portals
+	const uint portalID = globalInvocationID - u_FirstPortalGroup * 64;
+	if( globalGroupID >= u_FirstPortalGroup && ( portalID < u_TotalPortals ) ) {
+		const uint portalSurfaceID = portalID + ( u_Frame * MAX_VIEWS + u_ViewID ) * u_TotalPortals;
+		PortalSurface surface = portalSurfaces[portalSurfaceID];
+		bool culled = CullSurface( surface.boundingSphere );
+
+		portalSurfaces[portalSurfaceID].distance = !culled ? distance( u_CameraPosition, surface.boundingSphere.center ) : -1.0;
```

Comment:
If there are nested portals wouldn't the camera position need to be different?

Reply:
It would and it is, the correct position is passed to `QueueSurfaceCull()`.